### ISSUE TITLE
feat: hybrid vector search — semantic augmentation for search_concepts

### ISSUE THREAD

**Issue by aTigar** [aTigar/ontobi#47]
## Motivation

Benchmark analysis (CIIT 2026, Run 2 — `benchmark_20260405_220250`) reveals that ontobi's 5-tier lexical search engine has **near-zero recall on inferential and cross-domain queries** (avg 0.44 for inferential, 0.66 for cross-domain). These are queries where the user describes a concept functionally rather than by name.

### Failure examples

| Query ID | Query (paraphrased) | Expected | Retrieved | Recall |
|----------|-------------------|----------|-----------|--------|
| i03 | "privacy implications of deploying ML models" | Membership Inference, CIA Triad | Differential Privacy, Privacy by Design, Adversarial ML | 0.0 |
| i07 | "vulnerability classification systems" | CWE, MITRE ATT&CK, Threat Modeling | Vulnerability, CVSS (noise) | 0.0 |
| i08 | "controls to limit blast radius of injection attack" | Input Validation, Separation of Duties | Evasion Attack, Attack Tree, Backdoor Attack (noise) | 0.0 |
| i10 | "defences against crafted input perturbations" | Evasion Attack, Adversarial ML | (nothing — 0 files found in 6/9 runs) | 0.33 |
| i12 | "healthcare org stores patient records, runs ML inference" | CIA Triad, PII, Differential Privacy | Big Data concepts (noise) | 0.07 |

### Root cause

The lexical search (Tier 4 token matching) can only find concepts whose **labels or aliases** contain query tokens. When a user asks about "privacy implications of ML deployment", the tokens `[privacy, implications, deploying, machine, learning, models]` never match "Membership Inference" or "CIA Triad" in their labels. The definition-only bridging filter intentionally rejects concepts that only match in their definition text (to prevent false positives like "Data Lakehouse" matching "Data Lake" queries).

A semantic/vector search would directly solve this — "privacy implications of ML models" has high cosine similarity with Membership Inference's definition ("...determines whether a specific data point was part of a model's training set, raising significant privacy concerns...").

### Reuse existing embedder infrastructure

The benchmark already uses `BAAI/bge-base-en-v1.5` (768-dim) via `fastembed` for the qdrant MCP baselines. The same model and toolchain should be reused for ontobi's vector search layer. This avoids introducing a new embedding dependency and ensures the same semantic space.

### Architecture: parallel search with score-gated merge

Run the vector search **in parallel** with the existing 5-tier lexical search on every `search_concepts` call:

```
User query
    ├─> Lexical search (existing 5-tier engine) ──> lexical_results[]
    │                                                   each has: { id, label, match_tier, match_score }
    └─> Vector search (cosine similarity)        ──> vector_results[]
                                                        each has: { id, label, cosine_score }
                                                        
Merge:
  1. Compute score_threshold = min(cosine_score for any concept in lexical_results)
     (i.e., the lowest vector score among concepts that the lexical engine already found)
  2. If lexical_results is empty, use a configurable fallback threshold (e.g., 0.45)
  3. From vector_results, take all concepts with cosine_score >= score_threshold
     that are NOT already in lexical_results
  4. Return these as a separate field: `semantic_matches[]` in the response
```

This means:
- The lexical results are **unchanged** — no regression risk for queries that already work
- Semantic matches are clearly labeled so the LLM knows they're vector-based (different confidence level)
- The score gate is **adaptive** — it uses the lexical hits as a calibration anchor. If ontobi found good lexical matches at cosine 0.72, only semantic results ≥ 0.72 are included. If no lexical matches exist, a fallback threshold kicks in.

### Response shape change

```typescript
interface SearchConceptsResult {
  // Existing fields (unchanged)
  results: ConceptSummary[];      // lexical matches, scored and tiered as today
  
  // New field
  semantic_matches?: ConceptSummary[];  // vector matches above threshold, not in lexical results
                                        // match_tier = 6 (new tier), match_score = cosine_score
}
```

The LLM sees both sets and can decide which concepts to explore further via `expand_concept_graph` or `get_concept_content`.

### Embedding index

- **Corpus**: All concept `skos:prefLabel` + `skos:definition` concatenated (e.g., `"Membership Inference: An attack that determines whether a specific data point was..."`)
- **Size**: ~212 concepts, ~50KB text total → 212 vectors × 768 dims = ~650KB in memory
- **Index type**: Brute-force cosine similarity is fine for 212 vectors (sub-1ms). No need for HNSW at this scale.
- **Lifecycle**: Rebuild on startup from the SPARQL endpoint (same candidate pool fetch that Tier 4 already does). No persistence needed — the index is tiny and rebuilds in <1s.
- **Embedding runtime**: `fastembed` (same as qdrant ingestion) or `@xenova/transformers` (JS-native, avoids Python dependency). Given the MCP server is TypeScript, `@xenova/transformers` may be cleaner.

### Alternative: delegate to Qdrant

Instead of embedding in-process, the MCP server could query an existing Qdrant collection (the same one used for the benchmark baselines). This avoids any embedding at MCP startup:

1. On `search_concepts`, fire a parallel HTTP request to Qdrant's `/collections/{name}/points/search` endpoint with the query text
2. Qdrant returns top-k with cosine scores
3. Merge using the same score-gated logic above

**Pros**: Zero embedding code in the MCP server. Reuses the exact same vectors the benchmark validates against.
**Cons**: Runtime dependency on Qdrant being up. Tighter coupling. The MCP server would need the Qdrant URL and collection name as config.

Both approaches are viable — the in-process option is more self-contained, the Qdrant delegation is simpler but adds a dependency.

### MCP package changes (`packages/mcp/`)

1. **New file**: `src/search/vector.ts` — embedding index management (build from candidates, query, score-gated merge)
2. **Modified**: `src/tools/search-concepts.ts` — call vector search in parallel, merge results, add `semantic_matches` to response
3. **Modified**: `src/index.ts` — initialize embedding index on startup (after SPARQL is ready)
4. **New config**: `ONTOBI_EMBEDDING_MODEL` env var (default: `BAAI/bge-base-en-v1.5`), `ONTOBI_SEMANTIC_THRESHOLD` (default: `0.45`), optionally `ONTOBI_QDRANT_URL` if delegating
5. **New dependency**: `@xenova/transformers` or `fastembed` or `qdrant-client` (depending on approach)

### No changes to

- `ontobi-core` (Rust) — the SPARQL store is unchanged
- The 5-tier lexical engine — no regression risk
- `expand_concept_graph` or `get_concept_content`

## Expected Impact

Based on benchmark failure analysis, this should improve:
- **Inferential recall**: 0.44 → estimated 0.80+ (5 complete-failure queries become findable)
- **Cross-domain recall**: 0.66 → estimated 0.85+ (semantic bridging across ontology clusters)
- **Precision**: Unchanged or slightly improved (score gate prevents low-quality semantic noise)

## References

- Benchmark analysis: `research/ciit-2026/notebooks/01_full_benchmark_exploration.ipynb` (Sections 11-12)
- Ground truth: `ontobi-bench/queries/ground_truth.json` (65 queries, 6 difficulty levels)
- Qdrant ingestion script: `research/ciit-2026/scripts/ingest_qdrant.py` (embedding model config)
- Existing search tiers: `packages/mcp/src/search/tiers.ts`

**Comment by aTigar** [comment]
**Deferred to future work (Decision #27, CIIT 2026 paper).**

Post-Run 2 failure analysis confirmed that FM1+FM4 queries (semantic gap, no lexical path from query to concept) remain at 0.000 recall — exactly the gap hybrid vector search would close.

However, adding vector search to ontobi changes it from an *ontology-guided* system to an *ontology+vector hybrid*, which muddies the paper's core comparison (ontology vs. vector baselines). Both ontobi and the qdrant baselines would use the same embedding model (`BAAI/bge-base-en-v1.5`), making it unclear whether improvements come from the ontology structure or from combining signals.

The current paper narrative is stronger without it:
- Tier 4b (#49) already improved recall by +51% using pure lexical definition matching
- The FM1+FM4 gap at 0.000 is a clean, quantified limitation that motivates this issue as the natural next step
- 4–6 page IEEE format doesn't have room for an additional ablation axis

This issue remains the planned first feature for the post-paper version of ontobi.
